### ISSUE TITLE
FFI performance regression in Deno 1.33.0

### ISSUE THREAD

**Issue by milaz** [denoland/deno#18883]
Calls to (i32, i32) -> i32 functions became 22x slower.
See ffiRunA line.

Now:
```
Task ffi-bench deno --unstable bench --allow-ffi ./plugin-bench.js
cpu: Intel(R) Core(TM) i3-8130U CPU @ 2.20GHz
runtime: deno 1.33.0 (x86_64-unknown-linux-gnu)

file:./plugin-bench.js
benchmark      time (avg)             (min … max)       p75       p99      p995
------------------------------------------------- -----------------------------
ffiRunA    108.08 ns/iter(17.26 ns … 321.71 ns) 113.74 ns 210.62 ns 286.8 ns
ffiRunB    10.41 ns/iter(9.74 ns … 44.83 ns) 10.04 ns 23.47 ns 41.1 ns
ffiRunC    11.67 ns/iter(11.21 ns … 45.37 ns) 11.22 ns 44.24 ns 44.83 ns
Native     5.61 ns/iter(5.02 ns … 25.23 ns) 5.61 ns 7.49 ns 11.6 ns
```

Before:
```
Task ffi-bench deno --unstable bench --allow-ffi ./plugin-bench.js
cpu: Intel(R) Core(TM) i3-8130U CPU @ 2.20GHz
runtime: deno 1.32.5 (x86_64-unknown-linux-gnu)

file:./plugin-bench.js
benchmark      time (avg)             (min … max)       p75       p99      p995
------------------------------------------------- -----------------------------
ffiRunA    4.91 ns/iter(4.73 ns … 64.38 ns) 4.73 ns 7.19 ns 16.95 ns
ffiRunB    10.21 ns/iter(9.47 ns … 46.2 ns) 9.78 ns 35.93 ns 38.91 ns
ffiRunC    10.87 ns/iter(10.62 ns … 52.56 ns) 10.63 ns 18.24 ns 23.3 ns
Native     5.55 ns/iter(5.34 ns … 30.03 ns) 5.35 ns 9.53 ns 27.14 ns
```

plugin.c
 
```
int a(int b, int c) {
    return b + c;
}

double b(double b, double c) {
    return b + c;
}

void c(const char* p) {
    return;
}
```

plugin-bench.js

```
let { symbols: { a, b, c } }  = Deno.dlopen('./plugin.so', {
    'a': {parameters: ['i32', 'i32'], result: 'i32'},
    'b': {parameters: ['f64', 'f64'], result: 'f64'},
    'c': {parameters: ['buffer'], result: 'void'},
});

let buf = new Uint8Array(10);

Deno.bench(function ffiRunA() { a(10, 20); });
Deno.bench(function ffiRunB() { b(2, 3); });
Deno.bench(function ffiRunC() { c(buf); });
Deno.bench(function Native() { return 2 + 3; });
```

**Comment by dsherret** [comment]
Does it change when you modify the order of the functions? (ex. when benching b before a)

**Comment by milaz** [comment]
Good idea. I have tested it now, and it really is the case. The first function, whatever it is, runs slower than the others.
It seems there is some warm up code introduced in 1.33.0 that causes that slowdown.

```
benchmark      time (avg)             (min … max)       p75       p99      p995
------------------------------------------------- -----------------------------
ffiRunB    116.6 ns/iter(18.36 ns … 348.05 ns) 115.57 ns 252.76 ns 286.62 ns
ffiRunA    9.83 ns/iter(8.71 ns … 48.2 ns) 9.16 ns 34.16 ns 43.12 ns
ffiRunC    12.37 ns/iter(11.21 ns … 49.31 ns) 11.22 ns 44.73 ns 44.85 ns
Native     5.83 ns/iter(5.03 ns … 58.73 ns) 5.62 ns 13.04 ns 23.14 ns
```

**Comment by milaz** [comment]
I did further investigation, adding noop as the first benchmark makes things to look right.
It seems that my initial guess reflected in the title of this issue is incorrect. This has nothing to do with FFI.
This is rather related to https://github.com/denoland/deno/issues/15277

**Comment by bartlomieju** [comment]
@milaz I already fixed the problem with benchmarking tool, it will be released in v1.35.2 later this week. Should we close the issue for now?

**Comment by milaz** [comment]
Thank you. I think this issue should be closed.
